### PR TITLE
Added document explorer custom QA logic to the samples.

### PR DESCRIPTION
**Pull request by ScopatGames**
*Issue #, if available:*
https://github.com/aws-samples/generative-ai-cdk-constructs-samples/issues/93

*Description of changes:*
Created a variant of the Document Explorer sample that passes custom Lambda code to the [aws-qa-appsync-opensearch](https://github.com/awslabs/generative-ai-cdk-constructs/blob/main/src/patterns/gen-ai/aws-qa-appsync-opensearch/README.md). The custom lambda allows the QA construct to process a list of multiple documents. The custom lambda code resides inside the `lib/lambda` directory.

Summarization and Visual Q&A logic was not changed.

In addition to the custom lambda code, the StreamLit client app was updated to display multiple document selection.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: samples/document-explorer-custom-qa/client_app/assets/doc_explorer_diagram.png
Comment (scottschreckengaust) [inline samples/document-explorer-custom-qa/client_app/assets/doc_explorer_diagram.png:1]: Could you upload the source diagram or export the PNG with metadata please?

File: samples/document-explorer-custom-qa/README.md [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+![Backend Architecture Diagram](docs/images/architecture.png)
+
+## Folder Structure
+
+This Document Explorer codebase is organized into folders containing the ```frontend``` and ```backend``` infrastructure code. The frontend client app is built with [Streamlit](https://streamlit.io/) and is located in the ```client_app``` folder. The backend code lives in ```bin/document_explorer.ts``` and uses the AWS CDK resources defined in the ```lib``` folder.
+
+The key folders are:
+
+```
+samples/document_explorer
```
Comment (scottschreckengaust) [inline samples/document-explorer-custom-qa/README.md:32]:
```suggestion
samples/document-explorer-custom-qa
```

Comment (ScopatGames) [reply samples/document-explorer-custom-qa/README.md]: Updated https://github.com/aws-samples/generative-ai-cdk-constructs-samples/pull/94/commits/96ada6015a1097902c368903a8509c3e70150ce0

**Comment by krokoko** [comment]
QQ: do you have by any chance a list of changes that you added to this sample vs the base doc explorer ? Maybe the PR could be a subfolder with those changes and guidance in the main repo, as it would avoid the code duplication (additional maintenance)

**Comment by ScopatGames** [comment]
> QQ: do you have by any chance a list of changes that you added to this sample vs the base doc explorer ? Maybe the PR could be a subfolder with those changes and guidance in the main repo, as it would avoid the code duplication (additional maintenance)

This may be suitable as a subfolder, though it requires small changes across the API Stack, importing the custom lambda code, and modifying the StreamLit app. Not sure the best way to do it would be. *Would simply adding this change to the Document Explorer sample be an option?*

Most of the diff shows as macro level file adds / deletions due to the changes to the directory structure, but here are the highlights:

1. *API Stack Change*: samples/document-explorer-custom-qa/lib/api-stack.ts Line 265: Added custom code to QaAppsyncOpensearch construct props.
2. *Gen AI QA Construct Lambda Logic copy*: samples/document-explorer-custom-qa/lib/lambda/aws-qa-appsync-opensearch-custom/*: Added (copied) lambda code from the QaAppsyncOpensearch default execution lambda.
3. *Lambda file that was changed*: samples/document-explorer-custom-qa/lib/lambda/aws-qa-appsync-opensearch-custom/question_answering/src/qa_agent/doc_qa.py Line 154: Contains the logic changes in the construct lambda to allow filtering of multiple documents.
4. *StreamLit modifications*: samples/document-explorer-custom-qa/client_app/pages/*: Updates to the Streamlit app to allow the front end to work with multiple document selections.

**Comment by krokoko** [comment]
> > QQ: do you have by any chance a list of changes that you added to this sample vs the base doc explorer ? Maybe the PR could be a subfolder with those changes and guidance in the main repo, as it would avoid the code duplication (additional maintenance)
> 
> This may be suitable as a subfolder, though it requires small changes across the API Stack, importing the custom lambda code, and modifying the StreamLit app. Not sure the best way to do it would be. _Would simply adding this change to the Document Explorer sample be an option?_
> 
> Most of the diff shows as macro level file adds / deletions due to the changes to the directory structure, but here are the highlights:
> 
> 1. _API Stack Change_: samples/document-explorer-custom-qa/lib/api-stack.ts Line 265: Added custom code to QaAppsyncOpensearch construct props.
> 2. _Gen AI QA Construct Lambda Logic copy_: samples/document-explorer-custom-qa/lib/lambda/aws-qa-appsync-opensearch-custom/*: Added (copied) lambda code from the QaAppsyncOpensearch default execution lambda.
> 3. _Lambda file that was changed_: samples/document-explorer-custom-qa/lib/lambda/aws-qa-appsync-opensearch-custom/question_answering/src/qa_agent/doc_qa.py Line 154: Contains the logic changes in the construct lambda to allow filtering of multiple documents.
> 4. _StreamLit modifications_: samples/document-explorer-custom-qa/client_app/pages/*: Updates to the Streamlit app to allow the front end to work with multiple document selections.

Thanks @ScopatGames ! Yes I think having it as a "guidance" in the doc explorer would be a better solution than a second doc_explorer. Maybe add a subfolder with a README explaining the changes to provide, and the lambda code you added there as well

**Comment by krokoko** [comment]
As discussed, closing this PR as another will be opened to have changes as a guidance under the original document explorer sample. Thanks @ScopatGames !

**Pull request closed without merging**